File: Android/PitchPerfect/src/depollsoft/pitchperfect/PitchPipeModel.java
package depollsoft.pitchperfect;

import depollsoft.lib.binding.ObservableCollection;
import depollsoft.lib.binding.TrackableField;
import depollsoft.lib.util.Preferences;
import depollsoft.pitchperfect.lib.Accidental;
import depollsoft.pitchperfect.lib.Note;

public class PitchPipeModel {
  private static final String PitchPipeModelKey = "depollsoft.pitchperfect.PitchPipeModel";
  private ObservableCollection<Note> cToC;
  private ObservableCollection<Note> fToF;

  private TrackableField<ObservableCollection<Note>> notes = new TrackableField<ObservableCollection<Note>>();

  public PitchPipeModel() {
    this.cToC = new ObservableCollection<Note>();
    this.cToC.add(Note.findNote("C", Accidental.Natural, 4));
    this.cToC.add(Note.findNote("C", Accidental.Sharp, 4));
    this.cToC.add(Note.findNote("D", Accidental.Natural, 4));
    this.cToC.add(Note.findNote("D", Accidental.Sharp, 4));
    this.cToC.add(Note.findNote("E", Accidental.Natural, 4));
    this.cToC.add(Note.findNote("F", Accidental.Natural, 4));
    this.cToC.add(Note.findNote("F", Accidental.Sharp, 4));
    this.cToC.add(Note.findNote("G", Accidental.Natural, 4));
    this.cToC.add(Note.findNote("G", Accidental.Sharp, 4));
    this.cToC.add(Note.findNote("A", Accidental.Natural, 4));
    this.cToC.add(Note.findNote("A", Accidental.Sharp, 4));
    this.cToC.add(Note.findNote("B", Accidental.Natural, 4));

    this.fToF = new ObservableCollection<Note>();
    this.fToF.add(Note.findNote("F", Accidental.Natural, 4));
    this.fToF.add(Note.findNote("F", Accidental.Sharp, 4));
    this.fToF.add(Note.findNote("G", Accidental.Natural, 4));
    this.fToF.add(Note.findNote("G", Accidental.Sharp, 4));
    this.fToF.add(Note.findNote("A", Accidental.Natural, 4));
    this.fToF.add(Note.findNote("A", Accidental.Sharp, 4));
    this.fToF.add(Note.findNote("B", Accidental.Natural, 4));
    this.fToF.add(Note.findNote("C", Accidental.Natural, 5));
    this.fToF.add(Note.findNote("C", Accidental.Sharp, 5));
    this.fToF.add(Note.findNote("D", Accidental.Natural, 5));
    this.fToF.add(Note.findNote("D", Accidental.Sharp, 5));
    this.fToF.add(Note.findNote("E", Accidental.Natural, 5));

    Preferences.initialize(PitchPipeModel.PitchPipeModelKey + ".IsFromFToF",
        false);
    if (this.getIsFromFToF())
      this.setNotes(this.fToF);
    else
      this.setNotes(this.cToC);
  }

  public boolean getIsFromFToF() {
    return Preferences.get(PitchPipeModel.PitchPipeModelKey + ".IsFromFToF");
  }

  public ObservableCollection<Note> getNotes() {
    return this.notes.getValue();
  }

  public void setIsFromFToF(boolean value) {
    Preferences.set(PitchPipeModel.PitchPipeModelKey + ".IsFromFToF", value);
    if (value)
      this.setNotes(this.fToF);
    else
      this.setNotes(this.cToC);
  }

  public void setNotes(ObservableCollection<Note> value) {
    this.notes.setValue(value);
  }
}

File: Android/PitchPerfect/src/main/java/depollsoft/pitchperfect/PitchPipeModel.kt
package depollsoft.pitchperfect

import depollsoft.lib.state.StateField
import depollsoft.lib.util.Preferences
import depollsoft.pitchperfect.lib.Accidental
import depollsoft.pitchperfect.lib.Note

/**
 * The pitch pipe's thirteen notes, C to C or F to F; the range is a stored preference.
 * [updateWidgets] redraws the home-screen widgets after the range changes.
 */
class PitchPipeModel @JvmOverloads constructor(
    private val updateWidgets: () -> Unit = PitchPipeAppWidget::updateWidgets,
) {
    /** The cells, clockwise from the top. */
    var notes: List<Note> by StateField(emptyList())

    /** Whether the pipe spans F4 to F5 rather than C4 to C5. Setting it redraws the widget too. */
    var isFromFToF: Boolean
        get() = Preferences.get(IS_FROM_F_TO_F_KEY)
        set(value) {
            Preferences.set(IS_FROM_F_TO_F_KEY, value)
            notes = if (value) F_TO_F else C_TO_C
            updateWidgets()
        }

    init {
        Preferences.initialize(IS_FROM_F_TO_F_KEY, false)
        notes = if (isFromFToF) F_TO_F else C_TO_C
    }

    companion object {
        private const val IS_FROM_F_TO_F_KEY = "depollsoft.pitchperfect.PitchPipeModel.IsFromFToF"

        private fun octave(
            vararg names: Pair<String, Int>,
            start: Int,
        ): List<Note> =
            names.map { (name, offset) ->
                val accidental = if (name.endsWith("#")) Accidental.Sharp else Accidental.Natural
                Note.findNote(name.removeSuffix("#"), accidental, start + offset)
            }

        private val C_TO_C: List<Note> by lazy {
            octave(
                "C" to 0, "C#" to 0, "D" to 0, "D#" to 0, "E" to 0, "F" to 0, "F#" to 0,
                "G" to 0, "G#" to 0, "A" to 0, "A#" to 0, "B" to 0, "C" to 1,
                start = 4,
            )
        }

        private val F_TO_F: List<Note> by lazy {
            octave(
                "F" to 0, "F#" to 0, "G" to 0, "G#" to 0, "A" to 0, "A#" to 0, "B" to 0,
                "C" to 1, "C#" to 1, "D" to 1, "D#" to 1, "E" to 1, "F" to 1,
                start = 4,
            )
        }
    }
}
